Profiler/Core/ProfilerPatch.cs

```diff
@@ -152,118 +175,147 @@ private static void SuffixProfilePb(ref ProfilerToken? __localProfilerHandle)
-                Log.Error(
-                    $"Didn't find any update profiling targets for {__methodBase.DeclaringType?.FullName}#{__methodBase.Name}.  Some profiling data will be missing");
+            {
+                Log.Error($"Didn't find any update profiling targets for {methodBaseName}.  Some profiling data will be missing");
+            }
+            
+            Log.Trace($"Finished TranspilerForUpdate for method {methodBaseName}");
 
             return il;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static ProfilerToken? Start(object obj)
+        {
+            switch (obj)
+            {
+                case MyEntityComponentBase componentBase:
+                {
+                    return new ProfilerToken(componentBase.Entity, GeneralEntrypoint, DateTime.UtcNow);
+                }
+                case IMyEntity entity:
+                {
+                    return new ProfilerToken(entity, GeneralEntrypoint, DateTime.UtcNow);
+                }
+                default:
+                {
+                    return null;
+                }
+            }
+        }
 
-        #region Distributed Update Targeting
-
-        private static bool IsDistributedIterate(MethodInfo info)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static ProfilerToken? StartProgrammableBlock(MyProgrammableBlock block)
         {
-            if (info == null)
-                return false;
-            if (!info.DeclaringType?.IsGenericType ?? true)
-                return false;
-            if (info.DeclaringType?.GetGenericTypeDefinition() != _distributedUpdaterIterate.DeclaringType)
-                return false;
-            var aps = _distributedUpdaterIterate.GetParameters();
-            var ops = info.GetParameters();
-            if (aps.Length != ops.Length)
-                return false;
-            for (var i = 0; i < aps.Length; i++)
-                if (aps[i].ParameterType.GetGenericTypeDefinition() != ops[i].ParameterType.GetGenericTypeDefinition())
-                    return false;
-            return true;
+            return new ProfilerToken(block, ScriptsEntrypoint, DateTime.UtcNow);
         }
 
-        private static void PatchDistributedUpdate(PatchContext ctx, MethodBase callerMethod)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static void StopToken(in ProfilerToken? tokenOrNull, bool mainThreadUpdate)
         {
-            var foundAnyIterate = false;
-            var msil = PatchUtilities.ReadInstructions(callerMethod).ToList();
-            for (var i = 0; i < msil.Count; i++)
+            if (!(tokenOrNull is ProfilerToken token)) return;
+
+            var result = new ProfilerResult(
+                token.GameEntity,
+                token.Entrypoint,
+                token.StartTimestamp,
+                DateTime.UtcNow,
+                mainThreadUpdate);
+
+            lock (_observers)
```

Comment:
StopToken is extremely hot code, and lock statements should rarely be in hot code.
Better to only ever update the observers list in Tick.  I believe Keen has a MyCachingList that queues up changes for later, then lets you apply them in bulk.